### PR TITLE
Handle files left open with immutable workspace mechanics on Windows

### LINKED ISSUE #27844: Temporary workspace cannot be moved to immutable location when files left open on Windows
### Current Behavior

See [google issue](https://issuetracker.google.com/issues/321877280) for more details, which google is looking at (I believe). Sync attempts on various projects that worked fine under Gradle 8.5 are failing using 8.6 RCx with this exception, which is being thrown by class `org.gradle.internal.execution.steps.AssignImmutableWorkspaceStep` which appears to be an internal Gradle exception. I can find no comments or Gradle doc that describe this exception or how to resolve it.  Android Studio Jellyfish is now requiring gradle 8.6.rc-1 or later, and this exception has been seen in both Gradle 8.6 rc-1 and rc-3.  Wanted to document this issue here in case it is a Gradle 8.6 issue before 8.6 goes production,

As indicated in the google issue doc, in many cases just retrying the sync works, as long as a new gradle daemon is not started between the error and the sync retry. Which seems unusual to me :-)

### Expected Behavior

Even if this exception is caused by some error in AGP and not Gradle 8.6, it seems like there should be doc or comments in the code or something that indicate possible causes of this Gradle exception, and even possible remedies. If this is a new bug in 8.6, then of course more than a doc change is required.

Any info/feedback on this would be appreciated by me, and maybe by whoever at Google the above issue is assigned to :-)

### Context (optional)

It can be avoided by not using 8.6, which means not using Android Studio Jellyfish. It can (at least in some cases) be worked around in Jellyfish and 8.6-rc-3 by doing additional sync retries until the sync completes without this exception. It is affecting project syncs on every project with which I have tried to use Android Studio Jellyfish.

### Steps to Reproduce

See [google issue](https://issuetracker.google.com/issues/321877280) 

### Gradle version

8.6 RC3

### Build scan URL (optional)

_No response_

### Your Environment (optional)

Android Studio Jellyfish Canary 5 using AGP 8.4.0-alpha05 running on Windows 11.

### PR DESCRIPTION
**Pull request by lptr**
Fixes #27844.

On Windows, immutable user work (typically non-incremental transforms) can leak file handles (i.e. leave files open), which prevents Windows from moving the temporary workspace to its immutable location, causing the related issue.

In this PR we handle this case better. When the move fails (for files left open or for another reason), we make a "defensive copy" of the workspace, where we control the state of file handles, and move that into the cache instead, then make a feeble attempt at removing the original temporary workspace, then log the problem.

### CODE REVIEW

**Comment by lptr** [comment]
@bot-gradle test QF please.

**Comment by lptr** [comment]
@bot-gradle test QF please.

**Comment by lptr** [comment]
@bot-gradle test this please.

File: subprojects/distributions-integ-tests/src/integTest/groovy/org/gradle/AllDistributionIntegrationSpec.groovy [hunk 1]
```diff
@@ -36,7 +36,7 @@ class AllDistributionIntegrationSpec extends DistributionIntegrationSpec {
 
     @Override
     int getMaxDistributionSizeBytes() {
-        return 212 * 1024 * 1024
+        return 215 * 1024 * 1024
```
Comment (lptr) [inline subprojects/distributions-integ-tests/src/integTest/groovy/org/gradle/AllDistributionIntegrationSpec.groovy:39]: 💭 These tests were broken, but for some reason never ran on `release`. @blindpirate is investigating.

**Comment by lptr** [comment]
@bot-grade test this please.

File: subprojects/distributions-integ-tests/src/integTest/groovy/org/gradle/AllDistributionIntegrationSpec.groovy (hunk 1 above)
Comment (lptr) [reply subprojects/distributions-integ-tests/src/integTest/groovy/org/gradle/AllDistributionIntegrationSpec.groovy]: Reverted this. The problem was caused by the long branch name being included in the zip over and over.

**Comment by lptr** [comment]
@bot-gradle test QF please.

**Review by asodja — APPROVED** [review]
LGTM

File: platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/AssignImmutableWorkspaceStep.java [hunk 2]
```diff
@@ -171,4 +159,100 @@ private static ImmutableListMultimap<String, HashCode> calculateOutputHashes(Imm
...
+        return failedMove.workspace.withTemporaryWorkspace(duplicateTemporaryWorkspace -> {
+            WorkspaceResult result = failedMove
+                .withDuplicatedTemporaryWorkspace(duplicateTemporaryWorkspace)
+                .executeMoveOrThrow();
+            failedMove.removeTemporaryWorkspace();
+            return result;
+        });
+    }
+
+    private class WorkspaceMoveHandler {
```
[the remark this answers is not in the document]
Comment (asodja) [reply platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/AssignImmutableWorkspaceStep.java:191]: Discussed already

[the remark this answers is not in the document]
Comment (lptr) [reply platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/AssignImmutableWorkspaceStep.java:191]: We'd need to keep the references to `deleter` and `fileSystemAccess` then. Let's keep it this way.

**Comment by lptr** [comment]
@bot-gradle test and merge please.

**Comment by lptr** [comment]
@bot-gradle merge this.

**Pull request merged**